parser/src/string.rs

```diff
@@ -0,0 +1,170 @@
+use crate::ast::{Constant, Expr, ExprKind, Location};
+use crate::error::LexicalErrorType::FStringError;
+use crate::error::{LexicalError, LexicalErrorType};
+use crate::fstring::parse_located_fstring;
+use crate::token::StringKind;
+use itertools::Itertools;
+
+pub fn parse_implicit_concatenation(
+    values: Vec<(Location, (String, StringKind))>,
+) -> Result<Expr, LexicalError> {
+    Ok({
+        // Preserve the initial location and kind.
+        let initial_location = values[0].0;
+        let initial_kind = (values[0].1 .1 == StringKind::U).then(|| "u".to_owned());
+
+        // Determine whether the list of values contains any f-strings. (If not, we can return a
+        // single Constant at the end, rather than a JoinedStr.)
+        let has_fstring = values
+            .iter()
+            .any(|(_, (_, string_kind))| *string_kind == StringKind::F);
+
+        // Collect all the sub-expressions: any Constants, along with any (flattened) f-strings.
+        let values: Vec<Expr> = values
+            .into_iter()
+            .map(|(location, (string, string_kind))| match string_kind {
+                StringKind::F => {
+                    if let ExprKind::JoinedStr { values } = parse_located_fstring(&string, location)
+                        .map_err(|e| LexicalError {
+                            location,
+                            error: FStringError(e.error),
+                        })?
+                        .node
+                    {
+                        Ok(values)
+                    } else {
+                        panic!("parse_located_fstring returned a non-JoinedStr.")
+                    }
+                }
+                _ => Ok(vec![Expr::new(
+                    location,
+                    ExprKind::Constant {
+                        value: Constant::Str(string),
+                        kind: (string_kind == StringKind::U).then(|| "u".to_owned()),
+                    },
+                )]),
+            })
+            .flat_map(|result| match result {
+                Ok(values) => values.into_iter().map(Ok).collect(),
+                Err(e) => vec![Err(e)],
+            })
+            .collect::<Result<Vec<Expr>, LexicalError>>()?;
+
+        // De-duplicate adjacent constants.
+        let mut deduped: Vec<Expr> = vec![];
+        let mut current: Vec<String> = vec![];
+        for value in values {
+            match value.node {
```

Comment:
match here is actually matching `StringKind::F` or not.
The entire upper match can be moved here.

Reply:
I was able to consolidate but I believe I still need an inner match?

An example would be something like:

```
(
  "foo"
  f"bar{1}baz"
)
```

In this case, we want to produce `JoinedStr([Constant("foobar"), FormattedValue(1), Constant("baz")])`. So I have to iterate over the values of f-string within the `StringKind` match, IIUC.